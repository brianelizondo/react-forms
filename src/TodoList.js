import React, { useState } from "react";
import { v4 as uuid } from 'uuid';
import "./BoxList.css";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import EditTodoForm from "./EditTodoForm";

function TodoList(){
    const INITIAL_STATE = [];
    const EDIT_INITIAL_STATE = { show: false, id: "", todo: "" };
    const [todos, setTodos] = useState(INITIAL_STATE);
    const [editTodo, setEditTodo] = useState(EDIT_INITIAL_STATE);

    const showEditForm = task => {
        setEditTodo({
            show: true,
            ...task
        });
    };

    const addTask = task => {        
        let newTask = { ...task, id: uuid(), completed: false };
        setTodos(tasks => [...tasks, newTask]);
    };
    const editTask = updateTask => {
        let updateTodos = todos.map(todo => {
            if(todo.id === updateTask.id){
                return { ...todo, task: updateTask.task }
            }
            return {...todo}
        });
        setTodos(tasks => updateTodos);
        setEditTodo(EDIT_INITIAL_STATE);
    };
    const completedTask = idTask => {
        let updateTodo = todos.map(todo => {
            if(todo.id === idTask){
                return { ...todo, completed: !todo.completed }
            }
            return {...todo}
        });
        setTodos(tasks => updateTodo);
    };
    const deleteTask = idTask => {
        setTodos(todos.filter(todo => todo.id !== idTask));
    };

    return (
        <div className="TodoList"> 
            <NewTodoForm addTask={addTask} />
            <div className="TodoList-container">
                { 
                    todos.map(todo => <Todo 
                        task={todo.task} 
                        id={todo.id} 
                        key={todo.id} 
                        showEditForm={showEditForm} 
                        completedTaskStatus={todo.completed} 
                        completedTask={completedTask} 
                        deleteTask={deleteTask} /> 
                    )
                }
                {editTodo.show && <EditTodoForm id={editTodo.id} task={editTodo.task} editTask={editTask} />}
            </div>
        </div>
    );
}

export default TodoList;