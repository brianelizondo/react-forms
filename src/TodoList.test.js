import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

// Smoke Tests
it("'TodoList' component renders without crashing", function() {
    render(<TodoList />);
});

// Snapshot Tests
it("'TodoList' matches snapshot", function() {
    const {asFragment} = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a new task", function() {
    const { getByLabelText, queryByText } = render(<TodoList />);
  
    // no todo yet
    expect(queryByText("X")).not.toBeInTheDocument();
  
    // form elements
    const taskInput = getByLabelText("Task:");
    const submitBtn = queryByText("Add Task!");
  
    // fill out the form
    fireEvent.change(taskInput, { target: { value: "test" }});
    fireEvent.click(submitBtn);
  
    // todo exists!
    expect(queryByText("X")).toBeInTheDocument();
  });