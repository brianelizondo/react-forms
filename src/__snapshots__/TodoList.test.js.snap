// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'TodoList' matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <div
      class="NewTodoForm"
    >
      <form>
        <label
          for="task"
        >
          Task:
        </label>
        <input
          id="task"
          name="task"
          placeholder="task"
          value=""
        />
        <button>
          Add Task!
        </button>
      </form>
    </div>
    <div
      class="TodoList-container"
    />
  </div>
</DocumentFragment>
`;
